### PR TITLE
Implement client api to generate and cache a random key

### PR DESCRIPTION
**Pull request by AlexLanzano**
- Add wh_Client_KeyCacheRandom client API to have the server generate a key from its RNG, cache it, and return only the keyId

- Add the client request/response/blocking triplet: wh_Client_KeyCacheRandomRequest_ex, wh_Client_KeyCacheRandomRequest, wh_Client_KeyCacheRandomResponse, and the blocking wh_Client_KeyCacheRandom

 - Add WH_KEY_CACHE_RANDOM keystore action and whMessageKeystore_CacheRandomRequest message (reusing CacheResponse) with its wh_MessageKeystore_TranslateCacheRandomRequest translation function

- Add server handler for WH_KEY_CACHE_RANDOM plus the _KeystoreCacheRandomKey helper, generating random bytes directly into a cache slot via wc_RNG_GenerateBlock

- Add whTest_KeyCacheRandom test (generate → export-verify size/label/non-zero → uniqueness → evict) wired into the crypto client test suite

- Document the new API in docs/

### CODE REVIEW

File: src/wh_server_keystore.c [hunk 1]
```diff
@@ -1694,6 +1733,56 @@ int wh_Server_HandleKeyRequest(whServerContext* server, uint16_t magic,
...
+                if (WH_KEYID_ISERASED(meta->id)) {
+                    ret = wh_Server_KeystoreGetUniqueId(server, &meta->id);
+                }
+                /* generate the key from the server RNG and cache it */
+                if (ret == WH_ERROR_OK) {
+                    ret = _KeystoreCacheRandomKey(server, meta);
+                }
+
+                (void)WH_SERVER_NVM_UNLOCK(server);
+            } /* WH_SERVER_NVM_LOCK() */
```
[the remark this answers is not in the document]
Comment (AlexLanzano) [reply src/wh_server_keystore.c]: Fixed by setting the sz field to 16 bit

File: wolfhsm/wh_client.h [hunk 2]
```diff
@@ -533,6 +533,78 @@ int wh_Client_KeyCache(whClientContext* c, uint32_t flags, uint8_t* label,
                        uint16_t labelSz, const uint8_t* in, uint16_t inSz,
                        uint16_t* keyId);
 
+/**
+ * @brief Sends a request to generate a key from the server RNG and cache it.
+ *
+ * The server generates keySz random bytes into a cache slot and returns only
+ * the assigned key ID; the raw key material never leaves the server. This
+ * function does not block; it returns immediately after sending the request.
```
[the remark this answers is not in the document]
Comment (AlexLanzano) [reply wolfhsm/wh_client.h]: Removed the statement

File: wolfhsm/wh_client.h [hunk 3]
```diff
@@ -533,6 +533,78 @@ int wh_Client_KeyCache(whClientContext* c, uint32_t flags, uint8_t* label,
...
+int wh_Client_KeyCacheRandomRequest(whClientContext* c, uint32_t flags,
+                                        uint8_t* label, uint16_t labelSz,
+                                        uint16_t keySz);
+
+/**
+ * @brief Receives a key generate response from the server.
+ *
+ * This function attempts to process the response to a key generate request. It
+ * validates the response and extracts the key ID. This function does not block;
+ * it returns WH_ERROR_NOTREADY if a response has not been received.
```
[the remark this answers is not in the document]
Comment (AlexLanzano) [reply wolfhsm/wh_client.h]: The implementation no longer reuses the wh_Client_KeyCacheResponse

File: src/wh_server_keystore.c [hunk 4]
```diff
@@ -850,6 +850,45 @@ int wh_Server_KeystoreCacheKeyChecked(whServerContext* server,
     return _KeystoreCacheKey(server, meta, in, 1);
 }
 
+#ifndef WC_NO_RNG
+/* Generate meta->len random bytes from the server RNG directly into a cache
+ * slot and commit it. No key material crosses the client/server boundary. */
+static int _KeystoreCacheRandomKey(whServerContext* server, whNvmMetadata* meta)
```
[the remark this answers is not in the document]
Comment (AlexLanzano) [reply src/wh_server_keystore.c]: Removed comment entirely

File: src/wh_server_keystore.c [hunk 5]
```diff
@@ -2272,6 +2309,68 @@ int wh_Server_HandleKeyRequest(whServerContext* server, uint16_t magic,
...
+                    req.labelSz = WH_NVM_LABEL_LEN;
+                }
+                memcpy(meta->label, req.label, req.labelSz);
+            }
+
+#ifndef WC_NO_RNG
+            if (ret == WH_ERROR_OK) {
+                ret = WH_SERVER_NVM_LOCK(server);
+            }
+            if (ret == WH_ERROR_OK) {
```
Comment (bigbrett) [inline src/wh_server_keystore.c]: style nit: nest lock body in the same if statement to preserve the "lock inspection at a glance" style like in the rest of the file

```
ret = LOCK()
if (ret == WH_ERROR_OK) {
    
    // do stuff

    (void)UNLOCK()
} /* LOCK() */
```

File: test/wh_test_crypto.c [hunk 6]
```diff
@@ -6753,6 +6753,90 @@ static int whTest_CacheExportKeyDma(whClientContext* ctx, whKeyId* inout_key_id,
...
+        keyId2 = WH_KEYID_ERASED;
+        ret    = wh_Client_KeyCacheRandom(ctx, 0, labelIn, sizeof(labelIn),
+                                              WH_TEST_KEYCACHERANDOM_KEYSIZE, &keyId2);
+        if (ret != 0) {
+            WH_ERROR_PRINT("Failed to wh_Client_KeyCacheRandom(2) %d\n", ret);
+        }
+        else if (keyId2 == keyId) {
+            WH_ERROR_PRINT("KeyCacheRandom reused keyId 0x%X\n", keyId);
+            ret = -1;
+        }
```
Comment (bigbrett) [inline test/wh_test_crypto.c:6830]: also check that random1 != random2?

File: wolfhsm/wh_client.h [hunk 7]
```diff
@@ -710,6 +710,78 @@ int wh_Client_KeyCache(whClientContext* c, uint32_t flags, uint8_t* label,
...
+ * @param[in] c Pointer to the client context.
+ * @param[in] flags Flags (whNvmFlags) for the generated key.
+ * @param[in] label Pointer to the label associated with the key.
+ * @param[in] labelSz Size of the label.
+ * @param[in] keySz Number of random key bytes to generate.
+ * @return int Returns 0 on success, or a negative error code on failure.
+ */
+int wh_Client_KeyCacheRandomRequest(whClientContext* c, uint32_t flags,
+                                        uint8_t* label, uint16_t labelSz,
+                                        uint16_t keySz);
```
Comment (bigbrett) [inline wolfhsm/wh_client.h]: I don't think we want/need two separate client functions for this, the EX version taking a keyId is fine. The client can supply keyId==ERASED to dynamically request one, since we already use this pattern in other functions.

**Review by bigbrett — CHANGES_REQUESTED** [review]
overall looks good, just a few tweaks

File: wolfhsm/wh_client.h (hunk 7 above)
Comment (AlexLanzano) [reply wolfhsm/wh_client.h]: Sounds good. Should I simply remove `wh_Client_KeyCacheRandomRequest`? or add the keyId to it and remove the _ex function?

File: src/wh_server_keystore.c (hunk 5 above)
Comment (AlexLanzano) [reply src/wh_server_keystore.c]: This clobbers the ret value above this. Would you be cool with this?

```
if (ret == WH_ERROR_OK) {
    ret = LOCK()
    if (ret == WH_ERROR_OK) {
    
        // do stuff

        (void)UNLOCK()
    } /* LOCK() */
}
```

File: wolfhsm/wh_client.h (hunk 7 above)
Comment (bigbrett) [reply wolfhsm/wh_client.h]: remove the _ex function and make the default API take a keyId

File: src/wh_server_keystore.c (hunk 5 above)
Comment (bigbrett) [reply src/wh_server_keystore.c]: yes that is  what I wanted you to do

Comment (AlexLanzano) [reply src/wh_server_keystore.c]: Fixed

File: test/wh_test_crypto.c (hunk 6 above)
Comment (AlexLanzano) [reply test/wh_test_crypto.c:6830]: Implemented this test

File: wolfhsm/wh_client.h (hunk 7 above)
Comment (AlexLanzano) [reply wolfhsm/wh_client.h]: Fixed

**Review by bigbrett — APPROVED** [review]

File: src/wh_server_keystore.c [hunk 8]
```diff
@@ -850,6 +850,43 @@ int wh_Server_KeystoreCacheKeyChecked(whServerContext* server,
...
+    }
+
+    /* Fill the slot directly from the server RNG */
+    ret = wc_RNG_GenerateBlock(server->crypto->rng, slotBuf, meta->len);
+    if (ret != 0) {
+        return ret;
+    }
+
+    memcpy((uint8_t*)slotMeta, (uint8_t*)meta, sizeof(whNvmMetadata));
+    _MarkKeyCommitted(_GetCacheContext(server, meta->id), meta->id, 0);
```
Comment (bigbrett) [inline src/wh_server_keystore.c:881]: @AlexLanzano Just noticed this. Why are we marking this as committed? Is it because we never want it to go to NVM and it is meant to be cache-only and auto-evictable? I'm wondering if we want to actually have this just be treated like any other key?

Comment (AlexLanzano) [reply src/wh_server_keystore.c:881]: Looks like Im setting it as uncommitted here, no? I used the same logic as the regular key cache handler

File: src/wh_client.c [hunk 9]
```diff
@@ -905,6 +905,98 @@ int wh_Client_KeyCache(whClientContext* c, uint32_t flags, uint8_t* label,
...
+        req->labelSz = capSz;
+        memcpy(req->label, label, capSz);
+    }
+
+    /* write request (no key material is sent) */
+    return wh_Client_SendRequest(c, WH_MESSAGE_GROUP_KEY, WH_KEY_CACHE_RANDOM,
+                                 sizeof(*req), (uint8_t*)req);
+}
+
+int wh_Client_KeyCacheRandomResponse(whClientContext* c, uint16_t* keyId)
```
Comment (bigbrett) [inline src/wh_client.c]: nit: name output pointer `outKeyId`

File: src/wh_client.c [hunk 10]
```diff
@@ -905,6 +905,98 @@ int wh_Client_KeyCache(whClientContext* c, uint32_t flags, uint8_t* label,
...
+            *keyId = resp->id;
+        }
+    }
+
+    return ret;
+}
+
+int wh_Client_KeyCacheRandom(whClientContext* c, uint32_t flags,
+                                 uint8_t* label, uint16_t labelSz,
+                                 uint16_t keySz, uint16_t* keyId)
```
Comment (bigbrett) [inline src/wh_client.c]: nit

```suggestion
                                 uint16_t keySz, uint16_t* inOutKeyId)
```

File: src/wh_server_keystore.c (hunk 8 above)
Comment (bigbrett) [reply src/wh_server_keystore.c:881]: facepalm.jpg

**Review by bigbrett — DISMISSED** [review]

File: src/wh_client.c (hunk 9 above)
Comment (AlexLanzano) [reply src/wh_client.c]: fixed

File: src/wh_client.c (hunk 10 above)
Comment (AlexLanzano) [reply src/wh_client.c]: fixed

**Review by bigbrett — APPROVED** [review]
